Add raw dependencies support to Android plugin configuration

belemaire opened the pull request:
Rather than introducing a new specific prefix for all type of dependencies statements as we started doing, this PR is adding a new `freeeform:` prefix for dependencies, to inject any arbitrary dependency string, allowing to cover all possible current and future needed dependency literal formats.

belemaire commented:
Dang 🤦 
While we're at it. 
Was just thinking about naming. Are you ok with `freeform:`. I was actually just thinking we could use something shorter such as `raw:` instead ? A preference here or something different ?

friederbluemle commented:
Good point! Yeah if I would need to choose between `raw:` and `freeform:` I'd probably pick `raw:`.

Another idea: What if we don't require a prefix at all and instead consider it raw/freeform _unless_ it falls into a category that is already supported? I think it can be done in a backwards compatible way. Something along the lines of:

1. Does it contain a known currently supported prefix (e.g. `transitive:`)? Process accordingly
2. Does it only contain two `:` and no spaces (for example)? Process it as a regular dependency
3. If not 1. and 2., consider it freeform/raw

Also need to update "as `compile` statements" to "as `implementation` statements` on line 155.

belemaire commented:
Yes that would be even better, thanks for the suggestion :+1:
Will have a take at it !

belemaire commented:
Ok done some refactoring to make it this way.
In addition to the `transitive:` prefix we also had support for `annotation:` and `files` prefixes.
Not sure where we were going with that, should have done "raw" support from the start not to bother adding special handling for all possible dependency statements case (there are a lot) would have made everyone lives simpler.
Anyway, deprecating all these prefixes ... I have removed mention of them from the documentation (well only `transitive:` was mentioned) and added deprecation warnings.
Will be able to remove them in a future release to clean things.
We now have only two cases now, either a string formatted as `<groupId>:<artifactId>:<version>` which will be injected as `implementation <groupId>:<artifactId>:<version>` and anything else, that will be injected raw.
Makes me wonder if we should even keep first format, but I'm OK with that so given that basic implementation statements are what is needed for 90% of dependencies, this just makes it shorter.

friederbluemle commented:
Awesome! :clap: Yeah, maybe good to keep for now for better backwards compatibility. We can eventually remove it during a bigger cleanup. :+1:
